src/drt/test/run-ispd.py

```diff
@@ -0,0 +1,173 @@
+            detailed_route -guide {bench_dir}/{design}/{design}.input.guide \\
+                           -output_guide {design}.output.guide.mod \\
+                           -output_maze {design}.output.maze.log \\
+                           -output_drc {design}.output.drc.rpt \\
+                           -verbose {verbose}
+            write_def {run_dir}/{design}/{design}.output.def
+            set drv_count [detailed_route_num_drvs] 
+            if {{ $drv_count > {drv} }} {{
+              puts \"ERROR: Increase in number of violations from {drv} to $drv_count\"
+              exit 1
+            }} elseif {{ $drv_count < {drv} }} {{
+              puts \"NOTICE: Decrease in number of violations from {drv} to $drv_count\"
+              exit 2
+            }}
+    """
+    with open(os.path.join(design_dir, "run.tcl"), 'w') as f:
+        print(textwrap.dedent(script), file=f)
+
+    print(f"Create run shell script for {design}")
+    run_sh = os.path.join(design_dir, "run.sh")
+    script = f"""
+            set -e
+            cd {design_dir}
+            {args.program} -exit run.tcl > {design_dir}/run_{design}.log
+            echo
+            echo
+            echo Report for {design}
+            tail {design_dir}/run_{design}.log
+            cd '{bench_dir}/ispd{ispd_year}eval'
+            ./ispd{ispd_year}eval \\
+                -lef {bench_dir}/{design}/{design}.input.lef \\
+                -def {design_dir}/{design}.output.def \\
+                -guide {bench_dir}/{design}/{design}.input.guide \\
+              | grep -v WARNING | grep -v ERROR
+            echo
+    """
+    with open(run_sh, 'w') as f:
+        print(textwrap.dedent(script), file=f)
+    st = os.stat(run_sh)
+    os.chmod(run_sh, st.st_mode | stat.S_IXUSR | stat.S_IXGRP)
+
+
+def test_enabled(design, patterns):
+    for pattern in patterns:
+        if fnmatch.fnmatch(design, pattern):
+            return True
+    return False
+    
+
+design_list_ispd18 = [
+    ("ispd18_test1",   0),
+    ("ispd18_test2",   0),
+    ("ispd18_test3",  14),
+    ("ispd18_test4",  13),
+    ("ispd18_test5",   0),
+    ("ispd18_test6",   0),
+    ("ispd18_test7",   0),
+    ("ispd18_test8",   0),
+    ("ispd18_test9",   0),
+    ("ispd18_test10",  0),
+]
+design_list_ispd19 = [
+    ("ispd19_test1",   0),
+    ("ispd19_test2",   0),
+    ("ispd19_test3",   0),
+    ("ispd19_test4",   0),
+    ("ispd19_test5",   0),
+    ("ispd19_test6",   3),
+    ("ispd19_test7",   0),
+    ("ispd19_test8",   1),
+    ("ispd19_test9",   0),
+    ("ispd19_test10", 19),
+]
+
+run_dir = os.path.join(args.dir, "runs")
+running_tests = set()
+for (design, drv) in design_list_ispd18:
+    if test_enabled(design, args.tests):
+        genFiles(run_dir, 18, design, drv)
+        running_tests.add(design)
+
+
+for (design, drv) in design_list_ispd19:
+    if test_enabled(design, args.tests):
+        genFiles(run_dir, 19, design, drv)
+        running_tests.add(design)
+
+os.chdir(run_dir)
```

Comment:
It would be better to not change to outside the current folder. Was a misguided approach on my part when I first wrote the Tcl script.

Reply:
I agree that would be good.

You'll have to update the detailed_route command to use full paths for its output.  Also {args.program} -exit run.tcl.